### PR TITLE
Add e2e test for metrics registration

### PR DESCRIPTION
**Pull request by sanchezl**

### CODE REVIEW

File: test/e2e/metrics_test.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "#") {
+			continue
+		}
+		entry := strings.SplitN(line, " ", 2)
+		metrics[entry[0]] = entry[1]
+	}
+	require.NoError(t, scanner.Err())
+	return metrics
```
Comment (sttts) [inline test/e2e/metrics_test.go]: make the test a bit easier to read with empty lines and high-level oneliner comments.

File: test/e2e/metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		t.Fatal("No metrics retrieved.")
+	}
+}
+
+func getMetricsForPod(t *testing.T, pod v1.Pod, kubeConfig *rest.Config, restClient *rest.RESTClient) map[string]string {
+	require.Equal(t, v1.PodRunning, pod.Status.Phase)
+	transport, upgrader, err := spdy.RoundTripperFor(kubeConfig)
+	require.NoError(t, err)
+	request := restClient.Post().Resource("pods").Namespace(pod.Namespace).Name(pod.Name).SubResource("portforward")
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, request.URL())
```
Comment (sttts) [inline test/e2e/metrics_test.go]: why do we need a port forwarder? This sounds super complex.

Comment (sttts) [reply test/e2e/metrics_test.go]: Not saying it is not necessary (maybe it is), but it surprises me as a reader and needs a comments explaining in 2 or 3 lines why.

File: test/e2e/metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
+package e2e
```
Comment (sttts) [inline test/e2e/metrics_test.go]: should this go to library-go as an e2e helper and just a oneliner here?

File: test/e2e/metrics_test.go (hunk 2 above)
Comment (sttts) [inline test/e2e/metrics_test.go]: didn't we move to http/2 by now?

File: test/e2e/metrics_test.go (hunk 3 above)
Comment (sanchezl) [reply test/e2e/metrics_test.go]: I intend to move to library-go once finalized.

File: test/e2e/metrics_test.go (hunk 2 above)
Comment (sanchezl) [reply test/e2e/metrics_test.go]: I've added some comments. AFAIK, port forwarding is needed in order for test to run outside of the cluster.

File: test/e2e/metrics_test.go (hunk 1 above)
Comment (sanchezl) [reply test/e2e/metrics_test.go]: Done.

File: test/e2e/metrics_test.go (hunk 3 above)
Comment (sanchezl) [reply test/e2e/metrics_test.go]: I've opened https://github.com/openshift/library-go/pull/596 to start.

**Comment by marun** [comment]
Why is it valuable to check that pods are exposing metric endpoints in the absence of checking that prometheus is correctly configured to consume output from those endpoints (e.g. https://github.com/openshift/origin/pull/24169)?  

It appears trivial to expose metric endpoints - any controller started by `NewControllerCommandConfig` gets `/metrics` for free. Wiring those endpoints properly to prometheus seems far more prone to misconfiguration and therefore deserving of testing effort, and such testing would implicitly guarantee the validation of the metrics endpoints.

**Comment by sanchezl** [comment]
> It appears trivial to expose metric endpoints - any controller started by `NewControllerCommandConfig` gets `/metrics` for free. ?  Wiring those endpoints properly to prometheus seems far more prone to misconfiguration and therefore deserving of testing effort, and such testing would implicitly guarantee the validation of the metrics endpoints.

This test is specifically about the /metrics endpoint being there, but not returning ANY metrics.

**Comment by sttts** [comment]
> but not returning ANY metrics.

I thought that was the regression we had: we had an empty /metrics endpoint, but it was always served. cc @mfojtik

**Comment by marun** [comment]
@sanchezl I have a [WIP test](https://github.com/openshift/service-ca-operator/pull/90/commits/2895b1f985f5ad2567fab7f4fd75fa0fb96a5ab1#diff-6413e1dbafd6dd63b5d907ea3ae6f6f9R861) that trivially checks that prometheus can scrape the endpoint for a namespace. Once the related PRs merge I intend to add it as a helper to library-go that will enable a one-line check that can be called to check metrics collection for this operator:

```golang
checkMetricsCollection(t, kubeConfig, operatorNamespace, [optional prometheus job name])
```

**Pull request closed without merging**